doc/designs/fips-indicator.md

```diff
@@ -0,0 +1,77 @@
+FIPS Indicator Design
+=====================
+
+This is a simple FIPS indicator API design based on a set of thread-local
+bitflags, which all start out set to 1. A bitflag is cleared if an operation is
+performed in violation of the requirements of an indicator.
+
+This enables ergonomic usage because it allows you to perform a sequence of
+cryptographic operations and then test if any of those operations were performed
+in violation of the requirements of FIPS at the end, rather than testing every
+single operation.
+
+```c
+/*
+ * The set of different indicators which can be reported.
+ */
+#define OSSL_FIPS_INDICATOR_FIPS_140_2      (1UL << 0)
+#define OSSL_FIPS_INDICATOR_FIPS_140_3      (1UL << 1)
+
+/*
+ * Retrieve the thread-local FIPS indicator flags. The returned value is a
+ * bitmask formed from OSSL_FIPS_INDICATOR values. Undefined bits (which have
+ * yet to be assigned) are 1 by default.
+ *
+ * If a bit is set to 1, this indicates that all operations occurring since that
+ * indicator bit was reset were performed in compliance with the requirements
+ * of the given indicator bit. If a bit is cleared to 0, this means that at
+ * least one operation in violation of the requirements of that indicator bit
+ * has been performed since the indicators were reset.
+ *
+ * The indicator flags are thread-local, and therefore reflect operations
+ * performed on the current thread.
+ */
+uint64_t OPENSSL_get_fips_indicators(void);
+
```

Comment:
Is a uint64_t going to be sufficiently large here? The granularity of our indicators may be such that we are going to have lots of them (several issues suggest that we may need an indicator for specific algorithms operating with specific modes/key sizes).  As such, as we move forward, we may well exceed 64 individual indicators